packages/presets/client/src/fragment-masking-plugin.ts

```diff
@@ -1,7 +1,7 @@
 import type { PluginFunction } from '@graphql-codegen/plugin-helpers';
 
 const fragmentTypeHelper = `
-export type FragmentType<TDocumentType extends DocumentNode<any, any>> = TDocumentType extends DocumentNode<
+export type FragmentType<TDocumentType extends DocumentTypeDecoration<any, any>> = TDocumentType extends DocumentTypeDecoration<
```

Comment:
Added a new type `DocumentTypeDecoration` in this PR: https://github.com/dotansimha/graphql-typed-document-node/pull/154. In here we don't really care if it's `DocumentNode` or `string` — the `__apiType?: (variables: TVariables) => TResult;` part is the important bit.

Reply:
I was even wondering whether we should use a helper type for this, but instead make it a spec for embedding type information in anything? https://github.com/dotansimha/graphql-typed-document-node/issues/143#issuecomment-1446106576 That would also make this kind of stuff more future-proof as adding a new generic argument would be a breaking change... (see https://github.com/dotansimha/graphql-typed-document-node/pull/148)

@beerose What do you think?

Reply:
I'm not quite sure if I get you, but we could remove the generic constraint here or default all type parameters in `DocumentTypeDecoration` to `any`, so we won't have a breaking change the next time we add something.

Alternatively, instead of a function signature in the phantom field, we could attach an object with any number of properties:

```ts
import type { DocumentNode } from "graphql"

interface TypeInfo {
  result?: unknown,
  variables?: unknown,
  value?: unknown
}

type DocumentTypeDecoration<TTypeInfo extends TypeInfo> = {
  __graphqlTypes?: TTypeInfo
}

export type TypedDocumentStringOperation<TVars, TRes> = string & DocumentTypeDecoration<{
  variables: TVars,
  result: TRes,
}>

export type ResultOf<T> = T extends DocumentTypeDecoration<infer TTypeInfo> ? TTypeInfo['result'] : never;

export type VariablesOf<T> = T extends DocumentTypeDecoration<infer TTypeInfo> ? TTypeInfo['variables'] : never;
```

then we'd be free to add any additional property:

```ts
interface TypedDocumentNodeFragment<T> extends DocumentNode, DocumentTypeDecoration<{
  value: T
}> {}
```

Reply:
[Playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAqgzhATgQQG4ENjsfJAxRdAcwFsIA7YKAXigG8oB9R0Sc9MgLigCJdEeAbijoMWHIgA23OMEQBLckSgBfKADJ6AKCi8oAEgBmhUhWAA5DhB4B+bnwQox2fgWJlKQrSsFatraH4AYWwAE1cTDypaAAo6HSYWcAore34vXXlQmTlFIl9ddi4oWQUlXzVNeN0eA2N3MwAlCEM4W24GfjRMF0c3U0puLuccPsizVQqASg16PSNxyksydt5gsIiGz0m-AKgAUQAPDjBJCCCAexIwC-IzAEUAVyQQGm1dZgCiiHsnl4yoI9HHYEroAD5QOKJL6pNaOWqVd66ZE1OqLYDNVqrTqOEKIcJjLbAIa4jaEgZUHyg5EqKbUiHkR6SSQVXxaAD07KgAFpeXz+QLBULhSLRWLubtklAAAqICDAYDyQwgAA8ABUAHxvapQADaAGkoIooABrCAgC6GKBqgC63DVBptFVmdCp-ilADV0JJnnAAPKGdVa2gOs0Wq22tl7GDkeS3NUXACSlCQCAAxorbiqYMHITAoBBDsAKKE4CJyK8bJCTUMZtQtagLlkoNw7qgkDNC8XyKXIQka0byIYkFBE1o6w2m6EElWx7pWxB24go1K8JJMN3+lE1ck4Cq8Lm8AWiyWywxagsiZi2nZB8PEFA8NfVDPH+uFRQt2Yd5A99TdLK8qKsqKr-sifokPIwD7gANPM9QUtetQQheCFRMs1hauoYEohBUGgSihHIjGcbkAmybFog6aZuQKpej6ED+oG5i3OYTLrgARmc+7XhqfEwThhGoeiGE8IJugamBkkotJLaPr4exrhuFB-IgIBfpQP6MSq9y5vcx7dr2DAGoOprmpaUD3Ha5avCor7GYaxphhZVncCx5Bscy6BcRA+7vpu6JaXu9yOnxL7zlAbZIL4nJQAABjGJDoHAJrNHATLAAGcVGnA5AAORUMAAAW0BgIgFyQA+8jXGcUSYKRUAAO7QI1FxMqEUBFaI0DGnFaaSPIZjcmVjHynFcHNXlzJQANZqSK8haQBmIjxVpoQACIXGmjzoRcoQQONUAcY8VDNSIcqAnAeQrVdNVKoNHVLldtydUg0DJU10ChLcBWdd1UDFRcCBGjVEB1TRZYlXKAB07qQLA5BJSlaUZQGQZvGqBmnnMwlXi0N7cOgFbhVAVZKR+5AacAQVBgk3Dk92qnqYFu5BmysXipzXPc5zkrwwzn7onAaqMcAACMbwC5TLO-tmpL4psFIajFXK6AAejYfhprcsgA6LUtU8Loti-T-mC0SNPrAr5JRLmOqfMk3xpPCAmZNkvBwFcEDclkPCu5FsI8J7ZDct8fsJKIPQSNIHte9ykfiNyjxSOHbqxQAsvIcBXUo8UJy4UjZYYg2SB1MTGFckWixAHVoWYdLa+QuvFrIBtCyLsgAEym8p0sW6zVsEvg6J2wksXImrMrlZVoBQHl+fR3lOVQJB2fXcaATQ1vCQO2wgfpP7WT2MH3u+-7TuxyHYcCW6ilm+QTMd+Lkv30zVM00cJxnJc1y3A8zxqWVhyVWUANZ80CIjZKqVGKo0MFpCWtBEpQJRpITKgZP6gx-jcO4lAmZAPHqAzWWhG7N1Fkg5GMDUEBhNgjJG0D0pUPQccTBVxsH-xeKPD4SQ97FB4EzcOuggRIA6NSXeKReEH2pEfS+p9QgCORBfIOcdr7UgXjAKQx844LyTinf2Kgb7sy5JnNeudvSSHikIxARcS6lmITrQqZDIEUIYWg7utDkGULQSqDBpxzisL-rggBIBOHQkdoHfh-tLEiJRLFSeWk54MDEYo9IwgVBLyzivLOOdlDFWgIYC4zILiNWumVCqSBFSMSgBXEgAMpR5USdw8RPw4RICEEad2pQ8jCAvp08oIgRjqJjr0-Iqg8rcCyHBb4cE1FSFEY05JLsEj6O8L4IAA)

Reply:
I meant more like that would be a breaking change for GraphQL Code Generator to support this. Because if a user is still using an older typed-document-node version, that has no 3rd generic parameter, and Code Generator would already generate code like that, it would result in TypeScript errors.

I think the function usage is there for a specific TypeScript behavior and quirk 🤔  https://github.com/dotansimha/graphql-typed-document-node/commit/6ee77c273422bea7df58f92607e54fe73eba673a